packages/next/types/index.d.ts

```diff
@@ -38,7 +38,7 @@ declare module 'react' {
  * `Page` type, use it as a guide to create `pages`.
  */
 export type NextPage<P = {}, IP = P> = {
-  (props: P): JSX.Element
+  (props: P & IP): JSX.Element
```

Comment:
I think we can refine this type a bit.

I can think of a few scenarios:
1. A Next.js page only rendered by Next (`NextPage<{ images: string[] }>`)
1. A homogeneous Next.js page rendered by Next or another Page (`NextPage<{ images: string[] }>`)
1. A heterogeneous Next.js page rendered with GIP by Next or props (and maybe partial GIP) by another Page (`NextPage<{ foo: boolean }, { images: string[] }>`)

It seems like we'd almost want this type to be `P` when `P === IP` but `Partial<IP> & P` when both are defined.

That means we can probably just write this type as `Partial<IP> & P`, assuming `(Partial<IP> & P) === P` when `IP === P`.

Note this would require we change the GIP signature too (so that `P` is always satisfied, users who want GIP-only would wrap their provided `P` as `Partial<{ ... }>`):
```ts
getInitialProps?(ctx: NextPageContext): Promise<IP & P>
```

----

What do you think, @maraisr?

Reply:
@maraisr This change won't work when using HOCs in the page that don't need to be returned by `getInitialProps`.
Consider this instead:
```tsx
getInitialProps?(ctx: NextPageContext): Promise<Partial<P> & IP>
```
Regardless of what you choose, add tests for it in `test/integration/typescript` 🙏 , for the tests you only need to add a component that uses the type in common scenarios.